README.MD

# NJUPT-API 简介

NJUPT-API 的初衷是希望为NJUPT的各个系统提供一套跨系统的简洁、优雅的、Pythonic的API接口，以便用户能够在此基础上进行扩展开发。

项目结构受到了开源项目 [zhihu-api](https://github.com/lzjun567/zhihu-api)的启发，在这里对作者[@lzjun567](https://github.com/lzjun567/)表示感谢

欢迎pr

# 目前实现的功能
## 正方系统

### 登录
```python
from njupt import Zhengfang
zhengfang = Zhengfang()
>> zhengfang.login(account='B1xxxxxxx',password='password')
    {
        'success':True,
        'code':0,
        'msg':'登录成功'
    }
# or zhengfang = Zhengfang('B1xxxxxxx','password')

```
### 获取课程成绩和绩点
```python
>>> zhengfang.get_score() 
    {'gpa': 4.99,
    'coursers': [{
        'year': '2015-2016', # 学年
        'semester': '1', # 学期
        'code': '00wk00003', # 课程代码
        'name': '从"愚昧"到"科学"-科学技术简史', # 课程名称
        'attribute': '任选', # 课程性质
        'belong': '全校任选课', # 课程归属
        'credit': '2.0', # 学分
        'point': '', # 绩点
        'score': '81', # 成绩
        'minor_mark': '0', # 辅修标记
        'make_up_score': '', # 辅修标记
        'retake_score': '', # 重修成绩 
        'college': '网络课程', # 开课学院
        'note': '', # 备注 
        'retake_mark': '0', # 重修标记
        'english_name': '' # 英文名称
        }, 
        ]
    }
```

### 获取等级考试信息
```python
>>> zhengfang.get_grade() 
    [
        {
        'date': '20151219',
         'name': '全国大学英语四级考试',
         'number': '320082152113313',
         'score': '547',
         'semester': '1',
         'year': '2015-2016'
        },
        ...
    ]
```
### 获取课表
```python
>>> zhengfang.get_schedule(week=1)
    # 二维列表，[i][j] 代表周i第j节课的课程。 为了方便，i或j为零0的单元均不使用。
    # 课程的节次使用一天12节课的形式描述。
    # 课程信息使用dict进行描述。
    # 一个参考的结构如下
    [
        [],
        [   None,
            {
            'classroom': '教4－202', 
            'name': '技术经济学', 
            'teacher': '储成祥'
            },
            ...
        ],
        ...
    ]



```

## 校园卡系统
### 登录
```python

from njupt import Card
>>> card = Card(account='11020xxxxxxxxxx',password='passwd')
    {
        'success': True,
        'code': 0, 
        'msg': '登录成功'
    }

# or card = Card(), card.login(account,password)
```

### 获取余额
```python
>>> card.get_balance()
    {
        'balance': 10.01,  # 到账余额
        'unsettle_balance': 0.01  # 过渡余额
        'total': 10.02  # 总余额
    }
```
### 充值(绑定银行卡 -> 校园卡)
```python
>>> card.recharge(amount=2.33)
    {   
        'success':True,  # 转账是否成功
        'code': 0,  # 状态码
        'msg': '转账成功'  # 附加信息
    }
```
### 获取账单
```python
>>> card.get_bill(start_date='2017-02-33',end_date='2018-01-03',rows=30,page=1)
    {'recodes': 
        [
            {'balances': 39.71, # 余额
              'change': -5, # 变动
              'comment': '未知系统,交电费', # 注释
              'department': '仙林售电处', # 操作部门
              'time': '2018-01-26 20:55:40', # 时间
              'type': '代扣代缴', # 类型
              'week': '星期五'}, # 星期
            {'balances': 39.71,
              'change': -7.5,
              'comment': '',
              'department': '一餐厅二楼清真食堂',
              'time': '2018-01-24 17:09:36',
              'type': '持卡人消费',
              'week': '星期三'},
               ... 
        ],
    'total': 52, # 总的记录数
    'total_pages':2,  # 总页数
    'page':1  # 当前的页码
    }
```

### 获取Dr.com的网费余额
```python
>>> card.get_net_balance()
    2.33
```

### 充值网费
```python
>>> card.recharge_net(amount=2.33)
    {
        'success': True, 
        'code' : 0,
        'Msg' : '充值成功'
    }
```
### 充值寝室电费
```python
>>> card.recharge_xianlin_elec(amount=2.33,building_name='兰苑11栋',room_id='4031')
    {
        'success': True, 
        'code' : 0,
        'Msg' : '缴费成功！'
    }
    # 三牌楼校区为card.recharge_sanpailou_elec()，参数相同（未测试）
```


# todos

图书馆系统和奥兰系统的常用接口

docs/changes.rst
Changelog
====================

Version 0.3.0
--------------

- 支持SSO登录正方 :class:`njupt.SSO` 支持利用my.njupt.edu.cn进行sso登录


Version 0.2.3
--------------

- 早锻炼系统 :class:`njupt.RunningMan` check时返回的打卡记录感知时区(UTC+8)


Version 0.2.2
--------------

- 支持早锻炼系统 :class:`njupt.RunningMan`


Version 0.1.5
--------------

- 支持图书馆 :class:`njupt.Library`
- 验证码识别重构，提高速度，减少依赖
- setup.py 使用 https://github.com/kennethreitz/setup.py 模板
- Pipfile不再指定Python版本

njupt/card.py
# encoding: utf-8
import base64
import datetime
import json
import re

from njupt.exceptions import NjuptException, AuthenticationException
from njupt.base import APIWrapper
from njupt.utils import login_required
from njupt.utils.captchas.card import CardCaptcha

AIDS = {
    'elec_xianlin': '0030000000005101',
    'elec_sanpailou': '0030000000005102',
    'net': '0030000000000301'
}


class Card(APIWrapper):
    """
    一卡通相关接口

    :param str username: 一卡通卡号
    :param str password: 一卡通查询密码

    :raise: :class:`njupt.exceptions.AuthenticationException`

    >>> card = Card(username='110201500111111', password='sssss')
    """

    class URLs:
        HOST = "http://yktapp.njupt.edu.cn:8070"
        # 登录地址
        LOGIN = HOST + '/Login/LoginBySnoQuery'
        # 一卡通验证码
        CAPTCHA = HOST + '/Login/GetValidateCode'
        # 一卡通基础信息
        INFO = HOST + '/User/GetCardInfoByAccountNoParm'
        # 一卡通消费记录
        BILL = HOST + '/Report/GetMyBill'
        # 一卡通充值
        RECHARGE = HOST + '/User/Account_Pay'
        # 一卡通Dr.com信息
        COMMON = HOST + '/Tsm/TsmCommon'
        # 一卡通网费充值
        NET_PAY = HOST + '/Tsm/Net_Pay'
        # 电费充值
        ELEC_PAY = HOST + '/Tsm/Elec_Pay'

    def __init__(self, username=None, password=None):
        super(Card, self).__init__()
        self.aid = None
        # 该系统使用aid区分业务类型
        self.inner_account = None
        self.verified = False
        if username and password:
            self.username = username
            self._login(username, password)

    def _login(self, account, password):
        captcha_code = CardCaptcha(self.get_image(self.URLs.CAPTCHA)).crack()
        data = {
            "sno": account,
            "pwd": base64.b64encode(str(password).encode("utf8")),
            "ValiCode": captcha_code,
            "remember": 1,
            "uclass": 1,
            "zqcode": "",
            "json": True,
        }

        result = self._login_execute(url=self.URLs.LOGIN, data=data)
        if result['code'] == 2:
            # 如果验证码错误，尝试递归重复登录
            return self._login(account, password)
        result['success'] = not result['code']
        if result['success']:
            self.verified = True
        else:
            raise AuthenticationException(result['msg'])
        return result

    def _login_execute(self, url=None, data=None):
        result = self.get_json(url=url, data=data, method="post")
        if result['IsSucceed']:
            return {'code': 0, 'msg': '登录成功'}
        else:
            if result['Msg'] == "验证码错误":
                return {'code': 2, 'msg': '验证码错误'}
            else:
                return {'code': 1, 'msg': result['Msg']}

    def _get_inner_account(self):
        """
        获取对应的内部账号，部分接口参数需要

        :return: 一卡通号对应的系统内部账号
        """
        if not self.inner_account:
            info = self.get_json(self.URLs.INFO, method='post', data={'json': True})
            result = json.loads(info['Msg'])
            self.inner_account = result['query_card']['card'][0]['account']
        return self.inner_account

    @login_required
    def get_balance(self):
        """
        查询余额

        :return: 包含 过度余额、到账余额、总余额的dict
        :rtype: dict of (str, float)

        >>> card.get_balance()
        {'balance': 10.01, 'unsettle_balance': 0.01, 'total': 10.02}
        """
        info = self.get_json(self.URLs.INFO, method='post', data={'json': True})
        result = json.loads(info['Msg'])
        balance = int(result['query_card']['card'][0]['db_balance']) / 100  # 到账余额
        unsettle_balance = int(result['query_card']['card'][0]['unsettle_amount']) / 100  # 过渡余额
        return {
            'balance': balance,  # 到账余额
            'unsettle_balance': unsettle_balance,  # 过渡余额
            'total': balance + unsettle_balance,  # 总余额
        }

    @login_required
    def get_net_balance(self):
        """
        获取Dr.com的上网费用余额

        :return: dr.com的余额
        :rtype: float

        >>> card.get_net_balance()
        2.33
        """
        data = {
            'jsondata': '{"query_net_info": {"aid": "%s", "account": "%s", "payacc": ""}}'  # 无法使用format格式化
                        % (AIDS['net'], self._get_inner_account()),
            'funname': 'synjones.onecard.query.net.info',
            'json': True,
        }
        result = self.get_json(url=self.URLs.COMMON, data=data, method='post')
        result = json.loads(result['Msg'])
        return float(re.search(r'余额(\d*\.\d*)元', result['query_net_info']['errmsg']).groups()[0])

    @login_required
    def recharge(self, amount):
        """
        从绑定的银行卡中扣款充值余额

        :param amount: 充值金额, 单位为元
        :type amount: float or int
        :return: 充值结果
        :rtype: dict

        >>> card.recharge(10)
        {'success': True, 'code': 0, 'msg': '转账成功'}
        """
        data = {
            'account': self._get_inner_account(),
            'acctype': '###',
            'tranamt': int(100 * amount),
            'qpwd': '',
            'paymethod': '2',
            'paytype': '使用绑定的默认账号',
            'client_type': 'web',
            'json': True,
        }
        info = self.get_json(self.URLs.RECHARGE, 'post', data)
        info = json.loads(info['Msg'])['transfer']
        result = {
            'code': int(info['retcode']),
            'msg': info['errmsg'],
            'success': not bool(int(info['retcode']))
        }
        return result

    @login_required
    def recharge_net(self, amount):
        """
        充值网费（一卡通余额 ==> 城市热点）

        :param amount: 充值金额, 单位为元
        :type amount: float or int
        :return: 充值结果

        >>> card.recharge_net(11)
        {'success': True, 'code' : 0, 'Msg' : '充值成功'}
        """
        data = {
            'paytype': 1,
            'aid': AIDS['net'],
            'account': self._get_inner_account(),
            'tran': int(amount * 100),
            'netacc': '{"netacc": "", "bal": null, "pkgid": null, "lostflag": null, "freezeflag": null, '
                      '"expflag": null,"statustime": null, "duration": null, "starttime": null, "pkgtab": []}',
            'pkgflag': 'none',
            'pkg': '{}',
            'acctype': '###',
            'qpwd': '',
            'json': True
        }
        result = self.get_json(url=self.URLs.NET_PAY, method='post', data=data)
        result = json.loads(result['Msg'])
        return {
            'success': not int(result['pay_net_gdc']['retcode']),
            'code': int(result['pay_net_gdc']['retcode']),
            'msg': result['pay_net_gdc']['errmsg']
        }

    def _get_build_ids(self, aid):
        """获取建筑的id"""
        data = {
            'jsondata': '{"query_elec_building": {"aid": "%s", "account": "%s", "area": {"area": "", "areaname": ""}}}'
                        % (aid, self._get_inner_account()),
            'funname': 'synjones.onecard.query.elec.building',
            'json': True,
        }
        result = self.get_json(data=data, url=self.URLs.COMMON, method='post')
        result = json.loads(result['Msg'])
        # - -!
        building_id = {}
        for build in result['query_elec_building']['buildingtab']:
            building_id[build['building']] = build['buildingid']
        return building_id

    @login_required
    def recharge_xianlin_elec(self, amount, building_name, big_room_id, small_room_id=0):
        """
        充值仙林校区的寝室电费

        :param float amount: 充值金额
        :param str building_name: 楼栋名称，例如 "兰苑11栋"
        :param big_room_id: 大寝寝室号
        :type big_room_id: int or str
        :param small_room_id: 小寝寝室号， 不传或传0则表示充值为大寝电费， 1、2、3则充值对应小寝空调电费
        :type small_room_id: int or str
        :return: 充值结果
        :raise: :class:`njupt.exceptions.NjuptException` 楼栋名称不正确

        >>> card.recharge_xianlin_elec(
        >>>    amount=11,
        >>>    building_name='兰苑11栋',
        >>>    big_room_id=403,
        >>>    small_room_id=1
        >>> )
        {'code': 0, 'msg': '缴费成功！', 'success': True}
        """
        try:
            buiding_id = self._get_build_ids(AIDS['elec_xianlin'])[building_name]
            return self._recharge_electricity(amount=amount, elec_aid=AIDS['elec_xianlin'], building_id=buiding_id,
                                              building=building_name, room_id='{}{}'.format(big_room_id, small_room_id))
        except KeyError:
            raise NjuptException("不存在的楼栋")

    @login_required
    def recharge_sanpailou_elec(self, amount, building_name, room_id):
        """
        充值三牌楼校区的寝室电费，参数参考仙林校区（未测试
        """
        try:
            buiding_id = self._get_build_ids(AIDS['elec_sanpailou'])[building_name]
            return self._recharge_electricity(amount=amount, elec_aid=AIDS['elec_sanpailou'], building_id=buiding_id,
                                              building=building_name, room_id=room_id)
        except KeyError:
            raise NjuptException("不存在的楼栋")

    def _recharge_electricity(self, amount, elec_aid, building_id, building, room_id):
        data = {
            "acctype": "###",
            "paytype": 1,
            "aid": elec_aid,
            "account": self._get_inner_account(),
            "tran": int(amount * 100),
            "roomid": room_id,
            "room": "",
            "floorid": "",
            "floor": "",
            "buildingid": building_id,
            "building": building,
            "areaid": "",
            "areaname": "",
            "json": True
        }

        r = json.loads(self.get_json(self.URLs.ELEC_PAY, 'post', data=data)['Msg'])
        return {
            'success': not bool(int(r['pay_elec_gdc']['retcode'])),
            'code': int(r['pay_elec_gdc']['retcode']),
            'msg': r['pay_elec_gdc']['errmsg'],
        }

    @login_required
    def get_bill(self, start_date=(datetime.datetime.now() - datetime.timedelta(days=30)).strftime("%Y-%m-%d"),
                 end_date=datetime.datetime.now().strftime("%Y-%m-%d"), rows=100, page=1):
        """
        查询一卡通消费记录，默认为最近一个月的消费记录

        :param start_date: 如 "2017-12-27"
        :param end_date:  如 "2018-01-26"
        :param rows: 一次查询返回的最大记录数量,默认为100条记录
        :param page: 如果结果数量有多页，决定返回第几页。
        :return: 查询到的信息，格式见例子
        :rtype: dict

        >>> card.get_bill()
        {
            'recodes':
                [
                    {
                        'balance': 39.71,
                        'change': -5,
                        'comment': '未知系统,交电费',
                        'department': '仙林售电处',
                        'time': '2018-01-26 20:55:40',
                        'type': '代扣代缴',
                        'week': '星期五'
                    },
                    ...
                ],
            'total': 52
            'total_pages': 2
            'page': 1
        }
        """
        data = {
            "sdate": start_date,
            "edate": end_date,
            "account": self._get_inner_account(),
            "page": page,
            "rows": rows
        }
        temp = self.get_json(url=self.URLs.BILL, method="post", data=data)
        result = {'total': temp['total'], 'recodes': [], 'total_pages': temp['total'] // rows + 1, 'page': page}
        if temp['rows']:
            for row in temp['rows']:
                result['recodes'].append({
                    'balance': row['CARDBAL'],
                    'department': row['MERCNAME'].strip(),
                    'comment': row['JDESC'].strip(),
                    'change': row['TRANAMT'],
                    'week': row['XQ'],
                    'time': row['OCCTIME'],
                    'type': row['TRANNAME'].strip(),
                })
        return result
